### PR TITLE
Bug-fix: handle tx.to = 0x0 correctly

### PR DESCRIPTION
**Pull request by kunxian-xia**
### Issue Link

[issue#352](https://github.com/scroll-tech/zkevm-circuits/issues/352)

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: zkevm-circuits/src/tx_circuit.rs [hunk 1]
```diff
@@ -1932,4 +1932,21 @@ mod tx_circuit_tests {
...
+    #[test]
+    fn tx_circuit_to_is_zero() {
+        const MAX_TXS: usize = 1;
+        const MAX_CALLDATA: usize = 32;
+
+        let chain_id: u64 = mock::MOCK_CHAIN_ID.as_u64();
+        let mut tx = mock::CORRECT_MOCK_TXS[0].clone();
+        tx.to = Some(AddrOrWallet::from(address!(
+            "0x0000000000000000000000000000000000000000"
+        )));
```
Comment (roynalnaruto) [inline zkevm-circuits/src/tx_circuit.rs:1945]:
```suggestion
        tx.to = Some(AddrOrWallet::from(Address::zero()));
```

Comment (roynalnaruto) [reply zkevm-circuits/src/tx_circuit.rs:1945]: Not sure if `address!` macro was imported

**Review by roynalnaruto — APPROVED** [review]

File: zkevm-circuits/src/tx_circuit.rs (hunk 1 above)
Comment (kunxian-xia) [reply zkevm-circuits/src/tx_circuit.rs]: fixed.

**Review by roynalnaruto — COMMENTED** [review]
Had clicked "approve" by mistake earlier

File: zkevm-circuits/src/witness/tx.rs [hunk 2]
```diff
@@ -340,25 +345,23 @@ impl Encodable for SignedTransaction {
...
                 .from(mock_tx.from.address())
                 .nonce(mock_tx.nonce)
                 .gas_price(mock_tx.gas_price)
                 .gas(mock_tx.gas)
-                .to(callee_address)
                 .value(mock_tx.value)
                 .data(mock_tx.input.clone())
                 .chain_id(mock_tx.chain_id.as_u64());
+            if !is_create {
+                legacy_tx = legacy_tx.to(mock_tx.to.clone().map(|to| to.address()).unwrap());
```
Comment (roynalnaruto) [inline zkevm-circuits/src/witness/tx.rs:363]: is the `clone` required?

**Review by roynalnaruto — COMMENTED** [review]
Some nitpicks (feel free to ignore) and questions

File: zkevm-circuits/src/evm_circuit/execution/begin_tx.rs [hunk 3]
```diff
@@ -561,6 +561,7 @@ impl<F: Field> ExecutionGadget<F> for BeginTxGadget<F> {
             .assign(region, offset, caller_address)?;
         let callee_address = tx
             .callee_address
+            .unwrap_or(Address::zero())
```
[the remark this answers is not in the document]
Comment (kunxian-xia) [reply zkevm-circuits/src/evm_circuit/execution/begin_tx.rs:564]: I want to highlight that in the case that `tx.callee_address` is `None`, then the callee_address should be zero.

[the remark this answers is not in the document]
Comment (roynalnaruto) [reply zkevm-circuits/src/evm_circuit/execution/begin_tx.rs:564]: Yea, default case would be the zero address. But that's just a nitpick, no worries.

File: zkevm-circuits/src/witness/tx.rs [hunk 4]
```diff
@@ -296,10 +301,10 @@ impl Encodable for Transaction {
...
         s.append(&self.gas_price);
         s.append(&Word::from(self.gas));
-        if self.callee_address == Address::zero() {
-            s.append(&""); // address == 0, rlp = 0x80
+        if self.callee_address.is_none() {
+            s.append(&"");
         } else {
-            s.append(&self.callee_address);
+            s.append(&self.callee_address.unwrap());
         }
```
[the remark this answers is not in the document]
Comment (kunxian-xia) [reply zkevm-circuits/src/witness/tx.rs]: fixed in https://github.com/scroll-tech/zkevm-circuits/pull/353/commits/7fd8036b2d31f74846b95abc06c786dd00c6062d.

File: zkevm-circuits/src/witness/tx.rs [hunk 5]
```diff
@@ -324,10 +329,10 @@ impl Encodable for SignedTransaction {
         s.append(&Word::from(self.tx.nonce));
         s.append(&self.tx.gas_price);
         s.append(&Word::from(self.tx.gas));
-        if self.tx.callee_address == Address::zero() {
-            s.append(&""); // address == 0, rlp = 0x80
+        if self.tx.callee_address.is_none() {
```
[the remark this answers is not in the document]
Comment (kunxian-xia) [reply zkevm-circuits/src/witness/tx.rs]: fixed in https://github.com/scroll-tech/zkevm-circuits/pull/353/commits/7fd8036b2d31f74846b95abc06c786dd00c6062d

File: zkevm-circuits/src/witness/tx.rs (hunk 2 above)
Comment (kunxian-xia) [reply zkevm-circuits/src/witness/tx.rs]: removed in https://github.com/scroll-tech/zkevm-circuits/pull/353/commits/7fd8036b2d31f74846b95abc06c786dd00c6062d

**Pull request merged**